README.md
# Notion, but Cooler! 🌈

Host Notion on your own domain, with custom scripts & styling! (Not created or maintained by Notion.)

Last updated on May 8, see `CHANGELOG.md`. Pull requests welcome on [Github](https://github.com/pixelyunicorn/notion-but-cooler)!

**Disclaimer: This is a very hacky method that could break at any time.** Customize Notion at your own risk, and enjoy it while it still works!

## How to Setup Your Notion Page 👋

1. Remix this project!
1. Open up the Glitch editor, click the dropdown in the top left corner, and rename your project!
1. Open up server.js - this is where your configration goes.
1. Change `MY_DOMAIN` (server.js, line 18) to reflect your new project name (project-name.glitch.me).
1. In Notion, share your desired page with the public, and copy the page link.
1. Paste your public page link into `START_PAGE` (server.js, line 19)
1. Congrats! Your **Notion but Cooler 🌈** page is now live!

## What's Next? 🤔

In server.js, the `INJECT_INTO_HEAD` & `INJECT_INTO_FOOT` variables contain what gets injected into the HTML as the page loads.

- In `INJECT_INTO_HEAD`, you can include stylesheets, analytics scripts, meta tags and more.
- In `INJECT_INTO_HEAD`, there is also a section for meta tags to control what previews appear when the page is shared on chat apps and socail media. Make it your own by adding a title, description, social card, and favicon for your Notion page!
- In `INJECT_INTO_FOOT`, you can include javascript and other HTML code: this include analytics scripts, live chats, Glitch project details, and more.

I've also included some blank selectors in public/style.css to help get you started!

## Adding Custom Colours 🎨

Notion has a light mode (dark text on light bg) and a dark mode (light text on dark bg).

In the 5th/6th sections of public/style.css, you can edit the background colours of both modes to your liking.

Changing text colours consistently for the entire document is nearly impossible, however you can add text colours to individual items (like page titles).

(Note that `color` in CSS is spelled without a `u`, thanks America! 🇺🇸)

## Adding Google Fonts ✏️

1. Open up [fonts.google.com](https://fonts.google.com) and find a few fonts you like (for header, content, and monospace fonts).
1. Notion uses 400/500/700 weight (normal/book/bold) fonts. Add fonts with these weights if possible.
1. Copy your embed code that starts with `<link...`
1. Paste your embed code into `INJECT_INTO_HEAD` (server.js, line 26), replacing the existing line containing Google Fonts.
1. Update the top 4 sections of public/style.css with your desired fonts!

## Custom Domains 🌍

To add a custom domain on Glitch, you need to have 2 thanks from people. Alternatively, you can host this yourself!

1. Follow these steps: [glitch.com/help/custom-domain](https://glitch.com/help/custom-domain/)
1. Change MY_DOMAIN variable to reflect your new domain (server.js, line 17)
1. Wait up to 48 hours for changes to roll out

(PS If all you wanted to do was host Notion on a custom domain, without additional styling, you can clear the contents of public/style.css.)

## Limitations ⛔️

- This is a very hacky way of doing things that could break at any time.
- You may be able to access the Notion login or home but you can't actually login.
- Inline database titles may not follow your font settings.

## Examples 🖼

If you want inspiration, here are some styled Notion pages remixed from this project:

- Melody's Notion Page: [notion.melody.dev](https://notion.melody.dev) / [style.css](https://glitch.com/edit/#!/melody-notion?path=public%2Fstyle.css)

## Made by Melody ✨

The original project can be found at [glitch.com/~notion](https://glitch.com/~notion).

You can find me online at [melody.dev](https://melody.dev) or [@pixelyunicorn](https://twitter.com/pixelyunicorn).


CHANGELOG.md
## Changelog 🆕

- **Major Update #4 on May 8, 2020:**
  - server.js: Fix path of Notion JS files!
  - To upgrade, copy and paste server.js from line 73 and downwards!
- Minor Update #3 on May 4, 2020:
  - [Created a GitHub for pull requests!](https://github.com/pixelyunicorn/notion-but-cooler)
  - README.md: cleaned up documentation a bit
  - CHANGELOG.md: created changelog file
- Minor Update #2 on May 4, 2020:
  - style.css: added some more blank selectors for editor sidebars (for theming desktop app)
- Minor Update #1 on May 3, 2020
  - style.css: fixed code for frosted topbar on dark mode


server.js
// code below forked from this google doc for hosting on nodejs / glitch:
// https://docs.google.com/document/d/1pJ0Qeuqblp7NYQn3fhFttJKfXsdCuzkAoLnWqLBCjd8/edit

const express = require("express");
const events = require("events");
const fetch = require("node-fetch");
const app = express();
var eventEmitter = new events.EventEmitter();

// make all the files in 'public' available
app.use(express.static("public"));
app.use(express.json());

// set this up with a public notion page link and glitch domain
// (don't forget to name your project on glitch first!)
const MY_DOMAIN = "notion.glitch.me";
const START_PAGE =
  "https://www.notion.so/Custom-Notion-Page-518e7c416706412fba8dd5a3ef542ff5";

// code injected near the end of the <head> tag
// styling and meta tags should go here
let INJECT_INTO_HEAD = `
  <!-- google fonts -->
  <link href="https://fonts.googleapis.com/css2?family=Playfair+Display:ital,wght@0,400;0,700&family=Lato:ital,wght@0,400;0,500;0,700;1,400;1,700&family=Space+Mono:wght@400;700&display=swap" rel="stylesheet">

  <!-- stylesheet -->  
  <link rel="stylesheet" type="text/css" href="/style.css">
  
  <!-- meta tags -->
  <meta name="twitter:card" content="summary_large_image">
  <meta name="twitter:site" content="@NotionHQ">
  <meta name="twitter:title" content="Notion, But Cooler!">
  <meta name="twitter:description" content="A new tool that blends your everyday work apps into one. It's the all-in-one workspace for you and your team">
  <meta name="twitter:url" content="https://www.notion.so"><meta name="twitter:image" content="https://www.notion.so/images/meta/default.png">
  <meta property="og:site_name" content="Notion">
  <meta property="og:type" content="website">
  <meta property="og:url" content="https://www.notion.so">
  <meta property="og:title" content="Notion, But Cooler!">
  <meta property="og:description" content="A new tool that blends your everyday work apps into one. It's the all-in-one workspace for you and your team">
  <meta property="og:image" content="https://www.notion.so/images/meta/default.png">
  <meta property="og:locale" content="en_US">
  <link rel="shortcut icon" type="image/x-icon" href="/images/favicon.ico">
`;

// code injected near the end of the <body> tag
// analytics and other scripts should go here
let INJECT_INTO_FOOT = `
  <script src="/script.js"></script>
`;

eventEmitter.on("fetch", event => {
  event.respondWith(fetchAndApply(event.request));
});
const corsHeaders = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "GET, HEAD, POST,PUT, OPTIONS",
  "Access-Control-Allow-Headers": "Content-Type"
};
function handleOptions(request) {
  if (
    request.header("Origin") !== null &&
    request.header("Access-Control-Request-Method") !== null &&
    request.header("Access-Control-Request-Headers") !== null
  ) {
    // Handle CORS pre-flight request.
    response.set(corsHeaders);
  } else {
    // Handle standard OPTIONS request.
    response.append("Allow", "GET, HEAD, POST, PUT, OPTIONS");
  }
}

app.all("*", function(request, response) {
  async function fetchAndApply(request) {
    if (request.method === "OPTIONS") {
      return handleOptions(request);
    }
    let url = request.url;
    let res, body, contentType;
    console.log("fetch url: " + `https://www.notion.so${url}`);
    if (url.startsWith("/image")) {
      response.redirect(301, `https://www.notion.so${url}`);
    } else if (url.startsWith("/front")) {
      response.redirect(301, `https://www.notion.so${url}`);
    } else if (url.endsWith("js")) {
      response.type("application/x-javascript");
      res = await fetch(`https://www.notion.so${url}`);
      body = await res.text();
      try {
        body = body
          .replace(/www.notion.so/g, MY_DOMAIN)
          .replace(/notion.so/g, MY_DOMAIN);
        console.log("get rewrite app.js");
      } catch (err) {
        console.log(err);
      }
    } else if (url.endsWith("css")) {
      response.type("text/css");
      res = await fetch(`https://www.notion.so${url}`);
      body = await res.text();
    } else if (url.startsWith("/api")) {
      response.type("application/json;charset=UTF-8");
      response.append("Access-Control-Allow-Origin", "*");
      console.log("api method: " + request.method);
      res = await fetch(`https://www.notion.so${url}`, {
        body: JSON.stringify(request.body),
        headers: {
          "Content-Type": "application/json;charset=UTF-8",
          "User-Agent":
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.103 Safari/537.36"
        },
        method: request.method // *GET, POST, PUT, DELETE, etc.
      });
      body = await res.text();
    } else if (url === `/`) {
      let pageUrlList = START_PAGE.split("/");
      let redrictUrl = `https://${MY_DOMAIN}/${
        pageUrlList[pageUrlList.length - 1]
      }`;
      response.redirect(302, redrictUrl);
    } else {
      res = await fetch(`https://www.notion.so${url}`, {
        method: request.method // *GET, POST, PUT, DELETE, etc.
      });
      body = await res.text();
      contentType = await res.headers.get("content-type");
      try {
        response.type(contentType);
        if (contentType.startsWith("text/html")) {
          body = body.replace(/<\/head>/g, INJECT_INTO_HEAD + "</head>");
          body = body.replace(/<\/body>/g, INJECT_INTO_FOOT + "</body>");
        }
      } catch (err) {
        console.log(err);
      }
    }
    return body;
  }

  fetchAndApply(request).then(body => {
    response.send(body);
  });
});

// listen for requests :)
const listener = app.listen(process.env.PORT, () => {
  console.log("Your app is listening on port " + listener.address().port);
});
